### DISCUSSION TITLE
Allow declaring parameters without names in delegates

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6213] [Language Ideas]
Sometimes it might be useful to not declare a name for parameters in a declared delegate. Similar to how `Action<T1, T2, T3>` offers the very generic names "arg1", "arg2", "arg3", there is no parameter name declared. Some delegates serve simple purposes, or the parameter types are descriptive enough about what the parameter represents.

For example, consider the following delegate:
```csharp
public delegate IEnumerable<TEntry> QueryFunction(Table<TEntry> table);
```

The delegate only takes one parameter, which is profoundly the table on which the query is performed, as the type name suggests. Intuitively, it might come to a user to never define a parameter name, and get hit with an error saying that a parameter name is missing (speaking from experience).

What I suggest is being able to write the following (based on the code above):
```csharp
public delegate IEnumerable<TEntry> QueryFunction(Table<TEntry>);
```

The name of that parameter has little significance, and so is chosen to be omitted.

Under the hood, nameless delegate parameters will use the generic `argN` naming that is followed on `Func` and `Action` delegate types. In the above example, the parameter's name will be `arg1`, matching that of a `Func<T<TE>, IE<TE>>`.

**Comment by HaloFour** [comment]
Someone explicitly named those parameters in the `Action<...>` and `Func<...>` cases.  Even if the name isn't meaningful to you, it's still important and it's still a part of the public surface of that delegate and does impact code such as with invocation with named arguments.  Because of that any name synthesis would have to be predictable across any implementations of the language, effectively it'd have to be a part of the language specification.

**Reply by Rekkonnect** [reply]
If you notice, I specifically mention that the names on nameless parameters will be synthesized to match the currently used names on the most fundamental delegates. By "very generic", I mean to show that sometimes explicitly naming a delegate parameter doesn't help much, and could still function well without one.

**Reply by HaloFour** [reply]
> If you notice, I specifically mention that the names on nameless parameters will be synthesized to match the currently used names on the most fundamental delegates.

That's exactly what would have to be very explicitly part of the language specification.  The language team has generally shied away from doing stuff like this.

**Reply by Rekkonnect** [reply]
I don't get that last part; the language doesn't seem too hesitant to synthesize names, like it's being done on tuples, records, and so on.

And when name synthesis is avoided, it's because there is no way to control the name on the symbol whose name is synthesized. In this particular case, we're **enabling** name synthesis while the option to control the symbol name already exists.

**Reply by HaloFour** [reply]
@AlFasGD 

> I don't get that last part; the language doesn't seem too hesitant to synthesize names, like it's being done on tuples, records, and so on.

In both cases the names are well known and or provided by the developer.  They are never some default value, except for tuples, but the language isn't synthesizing the `Item1` names, those come from the underlying `ValueTuple<...>` types.

Yes, synthesizing `argN` names would be trivial to synthesize.  It's just as trivial for you to specify.
